internal/wazeroir/operations.go

```diff
@@ -1171,22 +1204,26 @@ func (OperationSet) Kind() OperationKind {
 	return OperationKindSet
 }
 
-// OperationGlobalGet implements Operation.
+//// OperationGlobalGet implements Operation.
+////
+//// The engines are expected to read the global value specified by OperationGlobalGet.Index,
+//// and push the copy of the value onto the stack.
+////
+//// See wasm.OpcodeGlobalGet.
+//type OperationGlobalGet struct{ Index uint32 }
 //
-// The engines are expected to read the global value specified by OperationGlobalGet.Index,
-// and push the copy of the value onto the stack.
+//// String implements fmt.Stringer.
+//func (o OperationGlobalGet) String() string {
+//	return fmt.Sprintf("%s %d", o.Kind(), o.Index)
+//}
 //
-// See wasm.OpcodeGlobalGet.
-type OperationGlobalGet struct{ Index uint32 }
-
-// String implements fmt.Stringer.
-func (o OperationGlobalGet) String() string {
-	return fmt.Sprintf("%s %d", o.Kind(), o.Index)
-}
+//// Kind implements Operation.Kind
+//func (OperationGlobalGet) Kind() OperationKind {
+//	return OperationKindGlobalGet
+//}
 
-// Kind implements Operation.Kind
-func (OperationGlobalGet) Kind() OperationKind {
-	return OperationKindGlobalGet
+func NewOperationGlobalGet(index uint32) OperationUnion {
+	return OperationUnion{OpKind: OperationKindGlobalGet, Us: []uint64{uint64(index)}}
```

Comment:
please do not use `Us []uint64` in general as much as possible, to avoid the allocation here. Instead, you can add `uint32` or `uint64` field in the union.

Reply:
sure, I didn't add new fields as I wanted it to be identical to `interfaceOp` so that we can unify easily at some point

https://github.com/tetratelabs/wazero/blob/main/internal/engine/interpreter/interpreter.go#L998-L1004

if we add a uint32 field I feel like _maybe_ we should add it to `interfaceOp` too? 🤔

Reply:
You won’t need interpreterOp any more with this union.

Reply:
You can delete the type

Reply:
what you are supposed to do is essentially lift entire lowerIr function (https://github.com/tetratelabs/wazero/blob/main/internal/engine/interpreter/interpreter.go#L290-L732) into wazeroir/compiler.go, and delete  `interpreterOp` and reuse the union everywhere in interpreter.

Reply:
right, I was trying to do it instruction-by-instruction because it is very disruptive for the compiler part, so my plan was to duplicate `interpreterOp` into a new struct, and then at some point delete the original one.

Reply:
it looks like we can use 2 uint64s in most places except

		case wazeroir.OperationV128Shuffle:
			op.us = make([]uint64, 16)
			for i, l := range o.Lanes {
				op.us[i] = uint64(l)
			}

Reply:
👍